plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/FilterToPinotSqlConverter.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.pinot.query;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.plugin.pinot.PinotColumnHandle;
+import io.trino.spi.connector.ColumnHandle;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.pinot.common.request.Expression;
+import org.apache.pinot.common.request.Function;
+import org.apache.pinot.common.request.Identifier;
+import org.apache.pinot.common.request.Literal;
+
+import java.math.BigDecimal;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+
+public class FilterToPinotSqlConverter
+{
+    private static final Map<String, String> BINARY_OPERATORS = ImmutableMap.<String, String>builder()
+            .put("equals", "=")
+            .put("not_equals", "!=")
+            .put("greater_than", ">")
+            .put("less_than", "<")
+            .put("greater_than_or_equal", ">=")
+            .put("less_than_or_equal", "<=")
+            .put("plus", "+")
+            .put("minus", "-")
+            .put("times", "*")
+            .put("divide", "/")
+            .build();
+
+    private FilterToPinotSqlConverter() {}
+
+    public static String convertFilter(org.apache.pinot.common.request.PinotQuery pinotQuery, Map<String, ColumnHandle> columnHandles)
+    {
+        StringBuilder builder = new StringBuilder();
+        new Visitor(builder, columnHandles).visitExpression(pinotQuery.getFilterExpression());
+        return builder.toString();
+    }
+
+    private static class Visitor
```

Comment:
This is not a "visitor" in the traditional sense of the word.

Also, having a stateful class with side-effects on the underlying StringBuilder is error prone and confusing. I would restructure it as either:
* a set of static methods that take a StringBuilder and append the formatted string to it
* or better, have each method return a String and let the caller compose them, similar to how ExpressionFormatter works. This is the easiest form to reason about. This will also allow for much simpler forms using String.format(), Joiner, etc.

Reply:
Updated, lmk, also thanks for the advice!